### PR TITLE
Remove hyper-tls and replace it with hyper-rustls

### PR DESCRIPTION
**Pull request by akoshelev**
Addresses #640

### CODE REVIEW

**Review by martinthomson — APPROVED** [review]
@andyleiserson is closer to this, but I can't see a downside here.  Especially given the removal of PKCS\#8 haxx for macs.

File: src/bin/helper.rs [hunk 1]
```diff
@@ -152,14 +102,9 @@ async fn server(args: ServerArgs) -> Result<(), BoxError> {
...
             let key = fs::read_to_string(&key_file).unwrap().into_bytes();
-            #[cfg(target_os = "macos")]
-            let key = pkcs8::munge_private_key(&key);
-
             (
-                ClientIdentity::Certificate(
-                    Identity::from_pkcs8(fs::read_to_string(&cert).unwrap().as_bytes(), &key)
-                        .unwrap(),
-                ),
+                ClientIdentity::from_pks8(fs::read_to_string(&cert).unwrap().as_bytes(), &key)
```
Comment (martinthomson) [inline src/bin/helper.rs:106]: Do you want to use `expect()` here to report errors?

File: src/net/client/mod.rs [hunk 2]
```diff
@@ -102,31 +133,46 @@ impl MpcHelperClient {
...
+                    store
+                };
+
+                let builder = builder.with_root_certificates(cert_store);
                 match identity {
-                    ClientIdentity::Certificate(identity) => {
-                        builder.identity(identity);
-                    }
+                    ClientIdentity::Certificate((cert, pk)) => builder
+                        .with_single_cert(vec![cert], pk)
```
Comment (martinthomson) [inline src/net/client/mod.rs:155]: I only just noticed this, but there is a potential risk here in that a PKCS\#8 file containing a certificate path (i.e., more than just the end-entity certificate) is not used.  I think that the right answer is to retain `Vec<Certificate>` instead.  The trust store in `cert_store` is fine (the end-entity certificate is all we need for that).

File: src/net/server/mod.rs [hunk 3]
```diff
@@ -533,6 +540,22 @@ mod e2e_tests {
...
+        fn verify_server_cert(
+            &self,
+            _end_entity: &Certificate,
+            _intermediates: &[Certificate],
+            _server_name: &ServerName,
+            _scts: &mut dyn Iterator<Item = &[u8]>,
+            _ocsp_response: &[u8],
+            _now: SystemTime,
+        ) -> Result<ServerCertVerified, rustls::Error> {
+            Ok(ServerCertVerified::assertion())
```
Comment (martinthomson) [inline src/net/server/mod.rs:555]: Don't we need dangerous_configuration enabled in rustls for this to work?

**Review by martinthomson — COMMENTED** [review]
A few more things to think about perhaps.

File: src/net/server/mod.rs (hunk 3 above)
Comment (akoshelev) [reply src/net/server/mod.rs:555]: that only brings a special `DangerousClientConfig` with `set_certificate_verifier` only. does not seem to be worth it

Comment (andyleiserson) [reply src/net/server/mod.rs:555]: `axum-server` enables it. (Not saying that is right or that we should rely on it, but that's what's going on.)

Comment (akoshelev) [reply src/net/server/mod.rs:555]: eh, true I missed that. let me be explicit about it in Cargo.toml

```
#[allow(unreachable_pub)]
#[derive(Debug)]
#[cfg_attr(docsrs, doc(cfg(feature = "dangerous_configuration")))]
pub struct ServerCertVerified(());
```

**Pull request merged**